### DISCUSSION TITLE
Validating the config key

### DISCUSSION THREAD

**Discussion by stetsche** [SafeExamBrowser/seb-mac#273] [Q&A] [answered]
Hi!
I am working on the integration of SEB for the Sakai LMS. The progress is going well, but I am having probelms validating the the config key hash. I am using the same hashing method as forthe exam key, which is working.
To find out if the issue lies in the input (SEB-JSON String) or the hashing method, I was trying to log the JSON value by enabling verbose logging (`<key>logLevel</key><integer>4</integer>`), but I didn't have success finding the value in the logs on macOS and Windows.
Here is an example JSON that I am generating based on the configuration, it is a pretty simple configuration, but is still fails:

`{"URLFilterEnable":false,"URLFilterEnableContentFilter":false,"URLFilterRules":[],"allowPreferencesWindow":false,"allowQuit":true,"allowSpellCheck":true,"allowWlan":true,"audioControlEnabled":true,"browserWindowAllowReload":true,"browserWindowWebView":3,"examSessionClearCookiesOnStart":false,"logLevel":4,"quitURL":"/quitSeb","showInputLanguage":true,"showReloadButton":true,"showTaskBar":true,"showTime":true,"startURL":"http://192.168.REDACTED:8080/samigo-app/servlet/Login?id=admin1674473216327"}`

This is how I generate the hash compared to the exam key hash
BEK : `hash(url + examKey)` comparing to received exam key hash
CK: `hash(url + hash(SEB-JSON)` comparing to received config key hash

If anyone has hints on how to log the JSON or anything that could be wrong related to the hashing, I would appreciate it a lot.

Best regards,
Markus Stetschnig

**Comment by danschlet** [comment, marked as answer]
Hi,
If you set logLevel to 4 (Verbose), you should find the JSON and the CK in one of the log files (you might have to search several, as verbose logging produces long logs which might get split up:

JSON for Config Key: 
Config Key: 

But I can already see the problem in your JSON: In SEB the alphabetic sorting of the keys in the JSON (unlike in settings Plist XML) is [case insensitive](https://safeexambrowser.org/developer/seb-config-key.html), so the URLFilter... keys should not be in the beginning of the JSON:

All <dict> elements from the plist XML must be ordered (alphabetically sorted) by their key names. Use a recursive method to apply ordering also to nested dictionaries contained in the root-level dictionary and in arrays. Use non-localized (culture invariant), ideally non-ASCII value based case insensitive ordering (non-ASCII value based case insensitive ordering: for example the key <key>allowWlan</key> comes before <key>allowWLAN</key>). Note: We removed the requirement for using non-ASCII value based case insensitive ordering, as APIs in operating systems often don't support non-ASCII value based case insensitive ordering reliably (it's undefined if lower or upper case letters are ordered higher). Instead same key names with different lower/uppercase letters should not be used.

**Comment by stetsche** [comment]
@danschlet
I didn't find the JSON in the logs, but you were right, the problem was the sorting, removed this properties to test and the hashes match. Thanks a lot!
